File: DSA/Assignment_6/decimal-binary.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Stack
{
    int top;
    int cap;
    int *arr;
}Stack;

Stack *createStack(int cap)
{
    Stack *stack=(Stack *)malloc(sizeof(Stack));
    stack->cap=cap;
    stack->top=-1;
    stack->arr=(int *)malloc(sizeof(int));
}

int isFull(Stack *stack)
{
    return stack->top==stack->cap-1;
}

int isEmpty(Stack *stack)
{
    return stack->top==-1;
}

void push(Stack *stack, int data)
{
    if(isFull(stack))
    {
        printf("Stack overflow!");
        return;
    }
    stack->arr[++stack->top]=data;
}

int pop(Stack *stack)
{
    if(isEmpty(stack))
    {
        printf("Stack underflow!");
        return 0;
    }
    return stack->arr[stack->top--];
}

int peek(Stack *stack)
{
    if(isEmpty(stack))
    {
        printf("Stack is empty!");
        return 0;
    }
    return stack->arr[stack->top];
}

int decimalToBinary(int num, Stack *stack)
{
    int rev_bi, bi;
    while(num%2==0)
    {
        rev_bi=rev_bi*10+(num%2);
        num/=2;
    }
    while(bi>0)
    {
        bi=bi*10+(bi%10);
        bi/=10;
    }
    return bi;
}

File: DSA/Assignment_6/decimal-binary.py
class Stack:
    def __init__(self, cap):
        self.arr=[]
        self.cap=cap
        self.top=-1

    @staticmethod
    def createStack(capacity):
        return Stack(capacity)
    
    def isFull(self):
        return self.top==self.cap-1
    
    def isEmpty(self):
        return self.top==-1
    
    def push(self, ele):
        if self.isFull():
            print("Stack Overflow!")
            return
        self.top+=1
        self.arr.append(ele)

    def pop(self):
        if self.isEmpty():
            print("Stack Underflow!")
            return
        self.top-=1
        return self.arr.pop()
    
    def peek(self):
        if self.isEmpty():
            print("Stack is Empty!")
            return
        return self.arr[self.top]
    
def decimalToBinary(num, stack):
    while num>0:
        stack.push(num%2)
        num/=2

def main():
    num=int(input("Enter a number:"))
    stack=Stack.creaeteStack()
    decimalToBinary(num)